src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskStatusUpdateProcessorImpl.scala

```diff
@@ -1,133 +1,85 @@
+    val taskId = status.getTaskId
+    val appId = taskIdUtil.appId(taskId)
 
-              taskOption match {
-                case Some(task) => postEvent(status, task)
-                case None       => log.warn(s"Task not found. Do not post event for '{}'", taskId.getValue)
-              }
-            }
-          }
+    val maybeTask = fetchTaskTimer(taskTracker.fetchTask(appId, taskId.getValue))
 
-        case TASK_RUNNING if !maybeTask.exists(_.hasStartedAt) => // staged, not running
-          taskTracker.running(appId, status)
-            .map { task => postEvent(status, task) }
-            .recover {
-              case NonFatal(t) =>
-                log.warn(s"Task could not be saved. Do not post event for '${taskId.getValue}'", t)
-                driverOpt.foreach(_.killTask(status.getTaskId))
-            }
-
-        case TASK_STAGING if !taskTracker.contains(appId) =>
-          log.warn(s"Received status update for unknown app $appId, killing task ${status.getTaskId}")
-          driverOpt.foreach(_.killTask(status.getTaskId))
-          Future.successful(())
+    val eventuallyMaybeApp = maybeTask match {
+      case Some(task) => getAppTimer(appRepository.app(appId, Timestamp(task.getVersion)))
+      case None       => Future.successful(None)
+    }
 
-        case _ =>
-          taskTracker.statusUpdate(appId, status).map {
-            case None =>
-              log.warn(s"Killing task ${status.getTaskId}")
-              driverOpt.foreach(_.killTask(status.getTaskId))
-            case _ =>
-          }
-      }
-      eventuallyTracked.map { _ =>
-        taskStatusEmitter.publish(incoming)
-        driverOpt.foreach(_.acknowledgeStatusUpdate(status))
-      }
+    eventuallyMaybeApp.flatMap { maybeApp =>
+      processUpdate(
+        timestamp = now,
+        appId = appId,
+        maybeApp = maybeApp,
+        maybeTask = maybeTask,
+        mesosStatus = status
+      )
+    }
   }
 
-  private[this] def rateLimit(appId: PathId, status: TaskStatus, taskOption: Option[MarathonTask]): Future[Unit] = {
-    taskOption match {
-      case Some(task) if status.getState != TaskState.TASK_KILLED =>
-        import scala.concurrent.ExecutionContext.Implicits.global
-        appRepository.app(appId, Timestamp(task.getVersion)).map { appOption =>
-          // It would be nice if we could make sure that the delay gets send
-          // to the AppTaskLauncherActor before we continue but that would require quite some work.
-          //
-          // In production, it should not matter if we do not delay some task launches so much.
-          appOption.foreach(launchQueue.addDelay)
+  private[this] def processUpdate(
+    timestamp: Timestamp,
+    appId: PathId,
+    maybeApp: Option[AppDefinition],
+    maybeTask: Option[MarathonTask],
+    mesosStatus: TaskStatus): Future[Unit] = {
+    steps.foldLeft(Future.successful(())) { (resultSoFar, nextStep) =>
+      stepTimers(nextStep.name).timeFuture {
```

Comment:
[Not for this PR] can we rename timeFuture into apply and handle it the same way as the non future based version?